numba/tests/test_heapq.py

```diff
@@ -0,0 +1,356 @@
+        yield a.tolist()
+        yield a[::-1].tolist()
+        self.rnd.shuffle(a)
+        yield a.tolist()
+
+    def test_heappush_basic(self):
+        pyfunc_push = heappush
+        cfunc_push = jit(nopython=True)(pyfunc_push)
+
+        pyfunc_pop = heappop
+        cfunc_pop = jit(nopython=True)(pyfunc_pop)
+
+        for iterable in self.iterables():
+            expected = sorted(iterable)
+            heap = [iterable.pop(0)]  # must initialise heap
+
+            for value in iterable:
+                cfunc_push(heap, value)
+
+            got = [cfunc_pop(heap) for _ in range(len(heap))]
+            self.assertPreciseEqual(expected, got)
+
+    def test_nsmallest_basic(self):
+        pyfunc = nsmallest
+        cfunc = jit(nopython=True)(pyfunc)
+
+        for iterable in self.iterables():
+            for n in range(-5, len(iterable) + 3):
+                expected = pyfunc(1, iterable)
+                got = cfunc(1, iterable)
+                self.assertPreciseEqual(expected, got)
+
+    def test_nlargest_basic(self):
+        pyfunc = nlargest
+        cfunc = jit(nopython=True)(pyfunc)
+
+        for iterable in self.iterables():
+            for n in range(-5, len(iterable) + 3):
+                expected = pyfunc(1, iterable)
+                got = cfunc(1, iterable)
+                self.assertPreciseEqual(expected, got)
+
+    def test_heapreplace_basic(self):
+        pyfunc = heapreplace
+        cfunc = jit(nopython=True)(pyfunc)
+
+        a = [1, 3, 5, 7, 9, 2, 4, 6, 8, 0]
+
+        heapify(a)
+        b = a[:]
+
+        for item in [-4, 4, 14]:
+            pyfunc(a, item)
+            cfunc(b, item)
+            self.assertPreciseEqual(a, b)
+
+        a = np.linspace(-3, 13, 20)
+        a[4] = np.nan
+        a[-1] = np.inf
+        a = a.tolist()
+
+        heapify(a)
+        b = a[:]
+
+        for item in [-4.0, 3.142, -np.inf, np.inf]:
+            pyfunc(a, item)
+            cfunc(b, item)
+            self.assertPreciseEqual(a, b)
+
+    def heapiter(self, heap):
+        try:
+            while 1:
+                yield heappop(heap)
+        except IndexError:
+            pass
+
+    def test_nbest(self):
+        # inspired by
+        # https://github.com/python/cpython/blob/e42b7051/Lib/test/test_heapq.py
+        pyfunc_heapify = heapify
+        cfunc_heapify = jit(nopython=True)(pyfunc_heapify)
+
+        pyfunc_heapreplace = heapreplace
+        cfunc_heapreplace = jit(nopython=True)(pyfunc_heapreplace)
+
+        data = self.rnd.choice(range(2000), 1000).tolist()
+        heap = data[:10]
+        cfunc_heapify(heap)
+
+        for item in data[10:]:
+            if item > heap[0]:
+                cfunc_heapreplace(heap, item)
+
+        self.assertPreciseEqual(list(self.heapiter(heap)), sorted(data)[-10:])
+
+    def test_heapsort(self):
+        # inspired by
+        # https://github.com/python/cpython/blob/e42b7051/Lib/test/test_heapq.py
+        pyfunc_heapify = heapify
```

Comment:
can probably just jit the function directly without the temporary variable (and below)?

Reply:
Done